app/modules/session/webauthn.server.ts

```diff
@@ -0,0 +1,145 @@
+
+  return options;
+}
+
+export async function verifyPasskeyRegistrationResponse(
+  user: UserWithAuthenticators,
+  response: RegistrationResponseJSON,
+): Promise<VerifiedRegistrationResponse> {
+  if (!user.currentChallenge) return { verified: false };
+
+  try {
+    const verification = await verifyRegistrationResponse({
+      response,
+      expectedChallenge: user.currentChallenge,
+      expectedOrigin: origin,
+      expectedRPID: rpID,
+    });
+
+    if (verification.verified) {
+      const { registrationInfo } = verification;
+      if (!registrationInfo) return { verified: false };
+
+      const { credentialPublicKey, credentialID, counter, credentialDeviceType, credentialBackedUp } = registrationInfo;
+
+      await db.authenticator.create({
+        data: {
+          userId: user.id,
+          credentialID: Buffer.from(credentialID)
+            .toString('base64')
+            .replace(/\+/g, '-')
+            .replace(/\//g, '_')
+            .replace(/=+$/, ''),
+          credentialPublicKey: Buffer.from(credentialPublicKey),
+          counter,
+          credentialDeviceType,
+          credentialBackedUp,
+        },
+      });
+      await db.user.update({ where: { id: user.id }, data: { currentChallenge: null } });
+    }
+
+    return verification;
+  } catch {
+    return { verified: false };
+  }
+}
+
+export async function getPasskeyAuthenticationOptions(user: UserWithAuthenticators) {
+  const options = await generateAuthenticationOptions({
+    rpID,
+    allowCredentials: user.authenticators.map((authenticator) => ({
+      type: 'public-key',
+      id: new Uint8Array(Buffer.from(authenticator.credentialID, 'base64')),
+    })),
+    userVerification: 'preferred',
+  });
+
+  await db.user.update({ where: { id: user.id }, data: { currentChallenge: options.challenge } });
+
+  return options;
+}
+
+export async function verifyPasskeyAuthenticationResponse(
+  user: UserWithAuthenticators,
+  response: AuthenticationResponseJSON,
+): Promise<VerifiedRegistrationResponse> {
+  const authenticator = user.authenticators.find(({ credentialID }) => {
+    return credentialID === response.id;
+  });
+  if (!authenticator || !user.currentChallenge) return { verified: false };
+
+  try {
+    const verification = await verifyAuthenticationResponse({
+      response,
+      expectedChallenge: user.currentChallenge,
+      expectedOrigin: origin,
+      expectedRPID: rpID,
+      authenticator: {
+        ...authenticator,
+        credentialID: new Uint8Array(Buffer.from(authenticator.credentialID, 'base64')),
+      },
+    });
+
+    if (verification.verified) {
+      const { authenticationInfo } = verification;
+      if (!authenticationInfo) return { verified: false };
+
+      const { newCounter } = authenticationInfo;
+
+      await db.authenticator.update({ where: { id: authenticator.id }, data: { counter: newCounter } });
```

Comment:
We could use Promise.all here or a DB transaction to do these two `await` in parallel!